### PR TITLE
Use binary search for vectorized_choice

### PR DESCRIPTION
**Pull request by zmbc**
### Description

- *Category*: performance
- *JIRA issue*: [MIC-3569](https://jira.ihme.washington.edu/browse/MIC-3569)
- *Research reference*: none

### Changes and notes

Note that I have made this drop-in replacement without diving too deep into the behavior of vectorized_choice. In particular, I find its handling of Pandas indexes quite confusing.

### Verification and Testing

Stepped through and verified that the old code and new code would get the same results for actual inputs.

This change has a relatively small impact on runtime with small populations, because in small populations the timestep runtime dominates the total runtime -- I am guessing because there is some constant overhead to each timestep, while population creation scales more exactly with population size.

Performance for a 250k population, before:

<pre><code>
2022-12-21 14:59:05.745 | vivarium.framework.engine:report:203 - 
              Event  Count  Mean time (s)  Std. dev. time (s)  Total time (s)  % Total time
     initialization      1           0.55                0.00            0.55          0.05
              setup      1           7.63                0.00            7.63          0.76
         post_setup      1           0.00                0.00            0.00          0.00
<b>population_creation      1         134.24                0.00          134.24         13.39</b>
 time_step__prepare    132           0.03                0.00            3.62          0.36
          time_step    132           6.21                0.21          819.96         81.78
 time_step__cleanup    132           0.21                0.02           27.36          2.73
    collect_metrics    132           0.06                0.10            7.87          0.78
     simulation_end      1           1.48                0.00            1.48          0.15
              total      1        1002.69                0.00         1002.69        100.00
</pre></code>

After:

<pre><code>
2022-12-21 15:17:57.076 | vivarium.framework.engine:report:203 - 
              Event  Count  Mean time (s)  Std. dev. time (s)  Total time (s)  % Total time
     initialization      1           0.97                0.00            0.97          0.11
              setup      1           7.29                0.00            7.29          0.80
         post_setup      1           0.00                0.00            0.00          0.00
<b>population_creation      1          18.35                0.00           18.35          2.01</b>
 time_step__prepare    132           0.03                0.01            3.77          0.41
          time_step    132           6.42                0.24          847.21         92.61
 time_step__cleanup    132           0.21                0.01           28.05          3.07
    collect_metrics    132           0.06                0.10            7.78          0.85
     simulation_end      1           1.44                0.00            1.44          0.16
              total      1         914.86                0.00          914.86        100.00
</pre></code>

Note that both of these runs included the change in #114.

I expect the speedup to become even more dramatic as we further increase population size, and as we expand from Florida to the full US for our input data.

### CODE REVIEW

File: src/vivarium_census_prl_synth_pop/utilities.py [hunk 1]
```diff
@@ -214,8 +214,7 @@ def vectorized_choice(
     cdf = np.cumsum(pmf)
 
     # for each p_i in probs, count how many elements of cdf for which p_i >= cdf_i
-    vect_find_index = np.vectorize(lambda p_i: (p_i >= cdf).sum())
-    chosen_indices = vect_find_index(probs)
+    chosen_indices = np.searchsorted(cdf[~pd.isnull(cdf)], probs, side="right")
```
Comment (rmudambi) [inline src/vivarium_census_prl_synth_pop/utilities.py:217]: When would `cdf` be null?

**Review by rmudambi — APPROVED** [review]

File: src/vivarium_census_prl_synth_pop/components/population.py [hunk 2]
```diff
@@ -205,7 +205,7 @@ def choose_group_quarters(self, target_number_sims: int) -> pd.Series:
             options=group_quarters,
             n_to_choose=target_number_sims,
             randomness_stream=self.randomness,
-            weights=self.population_data["households"][["person_weight"]],
+            weights=self.population_data["households"]["person_weight"],
```
Comment (albrja) [inline src/vivarium_census_prl_synth_pop/components/population.py:208]: Very curious how this wasn't breaking before since this would be a dataframe.

**Review by albrja — APPROVED** [review]

File: src/vivarium_census_prl_synth_pop/utilities.py (hunk 1 above)
Comment (zmbc) [reply src/vivarium_census_prl_synth_pop/utilities.py:217]: Sometimes the `weights` has NaNs in it (but non-NaN values for all the indices present in `options`). But, I am realizing this is only true of one call to `vectorized_choice` so it is probably easy to fix that one in this PR and remove this condition.

More broadly, the way this function interfaces between numpy and pandas seems suspect to me -- I think it depends on the index of `options` being sorted, when `options` is a Series -- but working that out and refactoring seemed potentially complex and out of scope for this performance improvement.

Note that none of this is new -- there were always NaNs in `cdf`, it was just hidden "better" by `(p_i >= cdf).sum()`.

File: src/vivarium_census_prl_synth_pop/components/population.py (hunk 2 above)
Comment (zmbc) [reply src/vivarium_census_prl_synth_pop/components/population.py]: To be honest, I am not sure. Something about using `vectorize` there.

Comment (zmbc) [reply src/vivarium_census_prl_synth_pop/components/population.py]: Or possibly `np.take` can transparently use a single-column dataframe, treating it like a 1d array.

Comment (zmbc) [reply src/vivarium_census_prl_synth_pop/components/population.py]: Ah, it was the first one: each call to the lambda inside `vectorize` returns a single-element pandas Series, but the vectorize function flattens each result into a simple integer data type (using the one value in the Series, and discarding the index) to return a numpy array.

**Pull request merged**